docs/reference/sql/data-definition.md

```diff
@@ -0,0 +1,228 @@
+) WITH (
+    kafka_topic = 'current_location',
+    partitions = 3,
+    value_format = 'json'
+);
+```
+
+As with a stream, you can declare a table directly on top of an existing
+{{ site.ak }} topic by omitting the number of partitions in the `WITH` clause.
+
+## Keys
+
+You can mark a column with the `KEY` keyword to indicate that it's a key
+column. Key columns constitute the key portion of the row's underlying
+{{ site.ak }} record. Only streams can mark columns as keys, and it's optional
+for them to do do. Tables must use the `PRIMARY KEY` constraint instead.
+
+In the following example statement, `k1`'s data is stored in the key portion of
+the row, and `v1`'s data is stored in the value.
+
+```sql
+CREATE STREAM s3 (
+    k1 VARCHAR KEY,
+    v1 VARCHAR
+) WITH (
+    kafka_topic = 's3',
+    value_format = 'json'
+);
+```
+
+The ability to declare key columns explicitly is especially useful when you're
+creating a stream over an existing topic. If ksqlDB can't infer what data is in
+the key of the underlying {{ site.ak }} record, it must perform a repartition
+of the rows internally. If you're not sure what data is in the key or you simply
+don't need it, you can omit the `KEY` keyword.
+
+## Default values
+
+If a column is declared in a schema, but no attribute is present in the
+underlying {{ site.ak }} record, the value for the row's column is populated as
+`null`.
+
+## Pseudocolumns
+
+A pseudocolumn is a column that's automatically populated by ksqlDB and contains
+meta-information that can be inferred about the row at creation time. By default,
+pseudocolumns aren't returned when selecting all columns with the star (`*`)
+special character. You must select them explicitly, as shown in the following
+example statement.
+
+```sql
+SELECT ROWTIME, * FROM s1 EMIT CHANGES;
+```
+
+The following table lists all pseudocolumns.
+
+| pseudocolumn | meaning                        |
+|--------------|--------------------------------|
+| `ROWTIME`    | Row timestamp, inferred from the underlying Kafka record if not overridden. |
+
+You can't create additional pseudocolumns beyond these.
+
+## Constraints
+
+Although data types help limit the range of values that can be accepted by
+ksqlDB, sometimes it's useful to have more sophisticated restrictions.
+_Constraints_ enable you to exercise this type of logic directly in your schema.
+
+### Primary key constraints
+
+In a relational database, a primary key indicates that a column will be used as
+a unique identifier for all rows in a table. If you have a table that has a row
+with primary key `5`, you can't insert another row whose primary key is also `5`.
+
+ksqlDB uses primary keys in a similar way, but there are a few differences,
+because ksqlDB is an event streaming database, not a relational database.
+
+- Only tables can have primary keys. Streams do not support them.
+- Adding multiple rows to a table with the same primary key doesn't cause the
+  subsequent rows to be rejected.
```

Comment:
This is mixing up the concept of a `table` with a `changelog`.

The table can only have one row with a specific primary key.  A changelog can contain multiple rows with the same key.  The changelog can be materialized into a table.

This is the same as a traditional database.

We should explain this table/changelog duality somewhere ;)

Yes, unfortunately at the moment we can `INSERT VALUES` into a table with an key for a row that already exists and it will work.  However, this is a bug IMHO, (it should be `UPSERT` not `INSERT`),.